python/pyarrow/tests/test_sparse_tensor.py

```diff
@@ -0,0 +1,221 @@
+        [0, 0, 0, 0, 0],
+        [0, 0, 0, 1, 0],
+        [0, 0, 0, 0, 0],
+        [0, 3, 0, 0, 0],
+    ])
+    dim_names = ['x', 'y']
+    sparse_tensor = sparse_tensor_type.from_dense_numpy(data, dim_names)
+
+    assert sparse_tensor.ndim == 2
+    assert sparse_tensor.size == 25
+    assert sparse_tensor.shape == data.shape
+    assert sparse_tensor.is_mutable
+    assert sparse_tensor.dim_name(0) == dim_names[0]
+    assert sparse_tensor.dim_names == dim_names
+    assert sparse_tensor.non_zero_length == 4
+
+
+def test_sparse_tensor_coo_base_object():
+    data = np.array([[4], [9], [7], [5]])
+    coords = np.array([[0, 0], [0, 2], [1, 1], [3, 3]])
+    array = np.array([[4, 0, 9, 0],
+                      [0, 7, 0, 0],
+                      [0, 0, 0, 0],
+                      [0, 0, 0, 5]])
+    sparse_tensor = pa.SparseTensorCOO.from_dense_numpy(array)
+    n = sys.getrefcount(sparse_tensor)
+    result_data, result_coords = sparse_tensor.to_numpy()
+    assert sys.getrefcount(sparse_tensor) == n + 2
+
+    sparse_tensor = None
+    assert np.array_equal(data, result_data)
+    assert np.array_equal(coords, result_coords)
+    assert result_coords.flags.f_contiguous  # column-major
+
+
+def test_sparse_tensor_csr_base_object():
+    data = np.array([[1], [2], [3], [4], [5], [6]])
+    indptr = np.array([0, 2, 3, 6])
+    indices = np.array([0, 2, 2, 0, 1, 2])
+    array = np.array([[1, 0, 2],
+                      [0, 0, 3],
+                      [4, 5, 6]])
+
+    sparse_tensor = pa.SparseTensorCSR.from_dense_numpy(array)
+    n = sys.getrefcount(sparse_tensor)
+    result_data, result_indptr, result_indices = sparse_tensor.to_numpy()
+    assert sys.getrefcount(sparse_tensor) == n + 3
+
+    sparse_tensor = None
+    assert np.array_equal(data, result_data)
+    assert np.array_equal(indptr, result_indptr)
+    assert np.array_equal(indices, result_indices)
+
+
+@pytest.mark.parametrize('sparse_tensor_type', [
+    pa.SparseTensorCSR,
+    pa.SparseTensorCOO,
+])
+def test_sparse_tensor_equals(sparse_tensor_type):
+    def eq(a, b):
+        assert a.equals(b)
+        assert a == b
+        assert not (a != b)
+
+    def ne(a, b):
+        assert not a.equals(b)
+        assert not (a == b)
+        assert a != b
+
+    data = np.random.randn(10, 6)[::, ::2]
+    sparse_tensor1 = sparse_tensor_type.from_dense_numpy(data)
+    sparse_tensor2 = sparse_tensor_type.from_dense_numpy(
+        np.ascontiguousarray(data))
+    eq(sparse_tensor1, sparse_tensor2)
+    data = data.copy()
+    data[9, 0] = 1.0
+    sparse_tensor2 = sparse_tensor_type.from_dense_numpy(
+        np.ascontiguousarray(data))
+    ne(sparse_tensor1, sparse_tensor2)
+
+
+@pytest.mark.parametrize('dtype_str,arrow_type', tensor_type_pairs)
+def test_sparse_tensor_coo_from_dense(dtype_str, arrow_type):
+    dtype = np.dtype(dtype_str)
+    data = np.array([[4], [9], [7], [5]]).astype(dtype)
+    coords = np.array([[0, 0], [0, 2], [1, 1], [3, 3]])
```

Comment:
@mrkn Can you quickly double-check this is ok? The original code had different values here.

Reply:
Results match [scipy's reference](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html). However our API for COO is slightly different to scipy's. Would it make sense to go closer to scipy?
Another reference is [pydata/sparse](https://sparse.pydata.org/en/latest/construct.html).

|                 | scipy                                              | pydata/sparse                       | pyarrow                                            |
|-----------------|----------------------------------------------------|-------------------------------------|----------------------------------------------------|
| coo constructor | row(1, M), col(1, M), data(M), shape(ndims)        | coords(2, M), data(M), shape(ndims) | coords(M, 2), data(M), shape(ndims)                |
| csr constructor | indptr(1, k), indices(1, M), data(M), shape(ndims) | NA                                  | indptr(1, k), indices(1, M), data(M), shape(ndims) |

Here `M` is number on non zero elements and `k` is number of pointers that will depend on the data present in the tensor.

Reply:
Well, the current API allows zero-copy between Numpy and Arrow, so I think it's fine.

Reply:
I just mean the 'python facing' interface, so users would see this:

`pyarrow`:
data = np.array([[4], [9], [7], [5]])
coords = np.array([[0, 0], [3, 3], [1, 1], [0, 2]])
pa.SparseTensorCOO.from_numpy(data, coords, (4, 4))

`scipy`:
row  = np.array([0, 3, 1, 0])
col  = np.array([0, 3, 1, 2])
data = np.array([4, 5, 7, 9])
coo_matrix((data, (row, col)), shape=(4, 4)).toarray()

Reply:
How is that different from currently?

Reply:
Well, `scipy` takes a tuple `(row, col)` of two `(1, M)` shaped ndarrays, while `pyarrow` would now take a single `coords` `(2, M)` shaped ndarray.

It's not a big difference but why introduce a new interface?

Reply:
Isn't the interface tied to the internal representation?

Reply:
Oh yeah, snap, sorry, I missed that. Ok, well no need to discuss `(1, M), (1, M)` vs `(M, 2)` here.

How about then going closer to `pydata/sparse` format? So `(M, 2)` vs `(2, M)`?
`pyarrow`:
```
pyarrow:
data = np.array([[4], [9], [7], [5]])
coords = np.array([[0, 0], [3, 3], [1, 1], [0, 2]])
pa.SparseTensorCOO.from_numpy(data, coords, (4, 4))
```
`pydata/sparse`:
```
coords = [[0, 1, 2, 3, 4],
          [0, 1, 2, 3, 4]]
data = [10, 20, 30, 40, 50]
s = sparse.COO(coords, data, shape=(5, 5))
```

Reply:
Well, either the PyArrow API is closer to `pydata/sparse` or it's closer to the C++ Arrow API.
In any case, it seems it's just a transpose?

Reply:
Yeah, and since transpose can be done from Python it's maybe best to leave this to integrations (https://issues.apache.org/jira/browse/ARROW-4224, https://issues.apache.org/jira/browse/ARROW-4223).

Reply:
(1) The reason why the current coords matrix is column-major is due to the discussion in #2546.
But I'm working on introducing the srides array in a SparseCOOIndex at https://github.com/mrkn/arrow/compare/c40d8e4...31beafd.
This allows us to have coords matrix in any-contiguousness.

(2) Arrow's COO sparse tensor manages an index as a matrix while SciPy's COO matrix manages an index in two vectors.  This difference comes from Arrow's sparse tensor is not a matrix, but a tensor.  Tensorflow and PyTorch employ the same approach for COO sparse tensor.

(3) If we should support zero-copy data sharing of a COO matrix with SciPy, we can introduce a special index format for COO matrix.

Reply:
Ok, I think this can be revisited later. Will merge.

Reply:
Great. Thanks @mrkn and @pitrou! :)